File: engine/src/main/java/com/routeforge/engine/geom/Haversine.java
package com.routeforge.engine.geom;

/**
 * Great-circle distance between two points on the Earth, in meters.
 * <p>
 * The Earth isn't a perfect sphere — but the haversine formula is plenty
 * accurate for routing (errors well under 0.5%). Used for:
 * <ul>
 *   <li>Edge lengths from OSM node coordinates.</li>
 *   <li>A*'s heuristic: a lower bound on the remaining distance to the goal.</li>
 * </ul>
 * The heuristic is <b>admissible</b> (never overestimates) because no path
 * along the road network can be shorter than the straight-line great-circle
 * distance — so A* with this heuristic always finds the true optimum.
 */
public final class Haversine {

    /** Mean Earth radius in meters (IUGG value). */
    public static final double EARTH_RADIUS_METERS = 6_371_008.8;

    private Haversine() { /* utility class */ }

    /**
     * Great-circle distance between two points, in meters.
     *
     * @param lat1 latitude of point 1 in degrees
     * @param lon1 longitude of point 1 in degrees
     * @param lat2 latitude of point 2 in degrees
     * @param lon2 longitude of point 2 in degrees
     */
    public static double distanceMeters(double lat1, double lon1, double lat2, double lon2) {
        double phi1 = Math.toRadians(lat1);
        double phi2 = Math.toRadians(lat2);
        double dPhi = Math.toRadians(lat2 - lat1);
        double dLam = Math.toRadians(lon2 - lon1);

        double sinHalfDPhi = Math.sin(dPhi / 2.0);
        double sinHalfDLam = Math.sin(dLam / 2.0);
        double a = sinHalfDPhi * sinHalfDPhi
                 + Math.cos(phi1) * Math.cos(phi2) * sinHalfDLam * sinHalfDLam;
        double c = 2.0 * Math.asin(Math.min(1.0, Math.sqrt(a)));
        return EARTH_RADIUS_METERS * c;
    }
}

File: engine-rs/core/src/haversine.rs
//! Great-circle distance on the sphere.
//!
//! Identical formula to the Java [`Haversine`] helper — kept in `f64`
//! throughout because the inputs are degrees that round-trip through
//! `sin`/`cos`, and `f32` rounding can introduce meter-scale error
//! over a one-kilometer edge.

const EARTH_RADIUS_METERS: f64 = 6_371_000.0;

#[inline]
pub fn distance_meters(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let phi1 = lat1.to_radians();
    let phi2 = lat2.to_radians();
    let dphi = (lat2 - lat1).to_radians();
    let dlam = (lon2 - lon1).to_radians();

    let a = (dphi * 0.5).sin().powi(2)
          + phi1.cos() * phi2.cos() * (dlam * 0.5).sin().powi(2);
    let c = 2.0 * a.sqrt().asin();
    EARTH_RADIUS_METERS * c
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zero_distance_when_points_match() {
        assert_eq!(distance_meters(47.15, 9.52, 47.15, 9.52), 0.0);
    }

    #[test]
    fn matches_known_distance() {
        // London (51.5074, -0.1278) -> Paris (48.8566, 2.3522) ~ 343 km.
        let d = distance_meters(51.5074, -0.1278, 48.8566, 2.3522);
        assert!((d - 343_500.0).abs() < 2_000.0, "got {d}");
    }
}
